### DISCUSSION TITLE
macOS proxy icon support and working directory

### DISCUSSION THREAD

**Discussion by dgl** [ghostty-org/ghostty#3987] [Feature Requests, Ideas]
Only supporting OSC 7 for the proxy icon is conflating the proxy icon with the working directory, which isn't always quite correct. Filing this is as a separate issue to #2681 but there is some overlap as to specs and what other terminals do.

Currently to get a document proxy icon, you have to use OSC 7 to set the path to a "document" (I [forked](https://github.com/dgl/vim-terminaldocument) vim-terminaldocument to make this work in Vim). However the result is when opening a new tab while vim is open, it has a cwd of `/`, because chdir fails[[1](https://github.com/ghostty-org/ghostty/blob/v1.0.0/src/Command.zig#L170)].

It would be possible to fix this by looking for a parent directory if the proxy icon is pointing to a file. However having OSC 6 and OSC 7 support would be better, for example consider a file manager, the "document" I'm looking at might be a directory several directories down from the cwd. Opening a new tab should go to the directory the file manager is in, not the open tab. Similar if you're using vim as an IDE, `:pwd` should be what a new tab opens in, but the proxy icon should allow me to drag a current document to a web browser, etc.

**Comment by mernen** [comment]
Agreed, it seems like a separate command is necessary in practice if we want an icon for working document. For example, when we close the editor we want to clear the working document but restore the icon for the working directory.

Terminal.app appears to use a simple rule: keep track of both working directory (OSC 7) and document (OSC 6), show proxy icon for document if it's not empty, directory otherwise. Opening a new window/tab on the same working directory is always based on OSC 7 alone.

**Comment by jcollie** [comment]
Is there any documentation on the format of OSC 6? The only thing I've been able to find so far is from [iTerm2](https://iterm2.com/documentation-escape-codes.html) and it does not use OSC 6 to keep track of the current document.

**Reply by j4james** [reply]
`OSC 6` is a bit of a mess, because it's been reused by multiple terminals for different purposes.

1. In Xterm it's used to enable and disable special color numbers, a synonym for `OSC 106` (documented on [the xterm ctlseqs page](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h4-Operating-System-Commands:OSC-Ps;Pt-ST:Ps-=-6;c;f.10FC)).
2. In Terminal.app and VTE, it's used to set the current document url, which I think is essentially what is being requested in this issue (documented in the Terminal.app preferences dialog, as shown in [this comment](https://github.com/alacritty/alacritty/pull/2937#issuecomment-557334660)).
3. In Eterm it was used to set various terminal properties, some of which are supported by iTerm2 (documented in the [Eterm Technical Reference](https://web.archive.org/web/20170204224835/http://eterm.org/docs/view.php?doc=ref#eterm_esc)).

I think iTerm2 actually supports option 2 as well, although I didn't see that in the iTerm2 documentation.
